applications/StructuralMechanicsApplication/custom_processes/metrics_spr_error_process.cpp

```diff
@@ -0,0 +1,659 @@
+        << "Error in percent: " << error_percentage << std::endl;
+
+    // Compute new element size
+    #pragma omp parallel for
+    for(int i_elem = 0; i_elem < num_elem; ++i_elem){
+        auto it_elem = elements_array.begin() + i_elem;
+
+        //Compute the current element size h
+        ComputeElementSize(it_elem);
+
+        // Compute new element size
+        const double element_error = it_elem->GetValue(ELEMENT_ERROR);
+        const double coeff = std::abs(element_error) < tolerance ? 1.0 : 1.0/element_error;
+        double new_element_size = coeff * it_elem->GetValue(ELEMENT_H);
+
+        // If a target number for elements is given: use this, else: use current element number
+        // if(mSetElementNumber == true && mElementNumber<mThisModelPart.Elements().size())
+        if(mSetElementNumber == true)
+            new_element_size *= std::sqrt((std::pow(rEnergyNormOverall, 2) + std::pow(rErrorOverall, 2))/mElementNumber) * mTargetError;
+        else
+            new_element_size *= std::sqrt((rEnergyNormOverall*rEnergyNormOverall+rErrorOverall*rErrorOverall)/mThisModelPart.Elements().size())*mTargetError;
+
+
+        // Check if element sizes are in specified limits. If not, set them to the limit case
+        if(new_element_size < mMinSize)
+            new_element_size = mMinSize;
+        if(new_element_size > mMaxSize)
+            new_element_size = mMaxSize;
+
+        it_elem->SetValue(ELEMENT_H, new_element_size);
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<SizeType TDim>
+void SPRMetricProcess<TDim>::CalculateMetric()
+{
+    const bool meshing_application_compiled = KratosComponents<Variable<Vector>>::Has("MMG_METRIC");
+    KRATOS_WARNING_IF("CalculateMetric", !meshing_application_compiled) << "Please compile the MeshingApplication in order to compute the metric" << std::endl;
+
+    // Actually computing the metric
+    if (meshing_application_compiled) {
+        // Getting metric variable
+        const Variable<Vector>& metric_variable = KratosComponents<Variable<Vector>>::Get("MMG_METRIC");
+
+        // Iteration over all nodes
+        NodesArrayType& nodes_array = mThisModelPart.Nodes();
+        const int num_nodes = static_cast<int>(nodes_array.size());
+
+        #pragma omp parallel for
+        for(int i_node = 0; i_node < num_nodes; ++i_node) {
+            auto it_node = nodes_array.begin() + i_node;
+            /**************************************************************************
+            ** Determine nodal element size h:
+            ** if mAverageNodalH == true : the nodal element size is averaged from the element size of neighboring elements
+            ** if mAverageNodalH == false: the nodal element size is the minimum element size from neighboring elements
+            */
+            double h_min = 0.0;
```

Comment:
This code block could be put into a separate function which takes a node & and returns h_min.